pkg/operator/certrotation/signer_test.go

```diff
@@ -19,6 +26,170 @@ import (
+	if err := indexer.Add(existing); err != nil {
+		t.Fatal(err)
+	}
+	clientset := kubefake.NewSimpleClientset(existing)
+
+	// the list cache is synced as soon as we have a delete, create, or update
+	clientset.PrependReactor("delete", "secrets", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+		if err = indexer.Delete(existing); err != nil {
+			t.Errorf("unexpected error while syncing the cache, op=delete: %v", err)
+		}
+		return false, nil, nil
+	})
+	clientset.PrependReactor("create", "secrets", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+		switch action := action.(type) {
+		case clienttesting.CreateActionImpl:
+			indexer.Delete(existing)
+			if err := indexer.Add(action.GetObject()); err != nil {
+				t.Errorf("unexpected error while syncing the cache, op=create: %v", err)
+			}
+			return false, action.GetObject(), nil
+		}
+		t.Errorf("wrong test setup")
+		return false, nil, nil
+	})
+	clientset.PrependReactor("update", "secrets", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+		switch action := action.(type) {
+		case clienttesting.UpdateActionImpl:
+			indexer.Delete(existing)
+			if err := indexer.Add(action.GetObject()); err != nil {
+				t.Errorf("unexpected error while syncing the cache, op=update: %v", err)
+			}
+			return false, action.GetObject(), nil
+		}
+		t.Errorf("wrong test setup")
+		return false, nil, nil
+	})
+
+	// r := rand.New(rand.NewSource(time.Now().Unix()))
+	options := events.RecommendedClusterSingletonCorrelatorOptions()
+	client := clientset.CoreV1().Secrets(ns)
+	var wg sync.WaitGroup
+	var id int32
+
+	controllers := 20
+	wg.Add(controllers)
+	for i := 1; i <= controllers; i++ {
+		go func() {
+			defer wg.Done()
+
+			controllerName := fmt.Sprintf("controller-%d", atomic.AddInt32(&id, 1))
+			recorder := events.NewKubeRecorderWithOptions(clientset.CoreV1().Events(ns), options, "operator", &corev1.ObjectReference{Name: controllerName, Namespace: ns})
+			wrapped := &wrapped{SecretInterface: client, name: controllerName, t: t}
+			getter := &getter{w: wrapped}
+			ctrl := &RotatedSigningCASecret{
+				Namespace:             ns,
+				Name:                  name,
+				Validity:              24 * time.Hour,
+				Refresh:               12 * time.Hour,
+				Client:                getter,
+				Lister:                corev1listers.NewSecretLister(indexer),
+				AdditionalAnnotations: AdditionalAnnotations{JiraComponent: "test"},
+				Owner:                 &metav1.OwnerReference{Name: "operator"},
+				EventRecorder:         recorder,
+			}
+
+			// <-time.After(time.Microsecond * time.Duration(r.Intn(100)))
+			_, err := ctrl.EnsureSigningCertKeyPair(context.TODO())
+			if err != nil {
+				t.Logf("error from controller 1 - %v", err)
+			}
+		}()
+	}
+
+	wg.Wait()
+	// controllers are done, we don't expect the signer to change
+	secretGot, err := client.Get(context.TODO(), name, metav1.GetOptions{})
```

Comment:
Is this test deterministic? I would rather see one or more tests like https://github.com/openshift/library-go/pull/1699 that always repro some of the known bad sequences, unless we can reproducibly fuzz different action sequences.